File: item-list/src/main/java/com/gurunars/item_list/ChangeOfPart.java
package com.gurunars.item_list;

import android.support.annotation.NonNull;

abstract class ChangeOfPart<ItemType extends Item> implements Change<ItemType> {

    int sourcePosition = -1;
    int targetPosition = -1;
    ItemType item;

    ChangeOfPart(ItemType item, int sourcePosition, int targetPosition) {
        this.item = item;
        this.sourcePosition = sourcePosition;
        this.targetPosition = targetPosition;
    }

    @Override
    public int hashCode() {
        int hash = getClass().getName().hashCode();
        return hash * sourcePosition +
                sourcePosition * targetPosition +
                hash * targetPosition;
    }

    @NonNull
    @Override
    public String toString() {
        return "new " + getClass().getName() + "<>(" + item.toString() + ", " + sourcePosition + ", " + targetPosition + ")";
    }

    private boolean samePositions(ChangeOfPart change) {
        return sourcePosition == change.sourcePosition && targetPosition == change.targetPosition;
    }

    @Override
    public boolean equals(Object object) {
        return object.getClass().equals(getClass()) && samePositions((ChangeOfPart) object);
    }

}

File: item-list/src/main/java/com/gurunars/item_list/ChangeOfPart.kt
package com.gurunars.item_list

internal abstract class ChangeOfPart<ItemType : Item>(var item: ItemType, sourcePosition: Int, targetPosition: Int) : Change<ItemType> {

    var sourcePosition = -1
    var targetPosition = -1

    init {
        this.sourcePosition = sourcePosition
        this.targetPosition = targetPosition
    }

    override fun hashCode(): Int {
        val hash = javaClass.name.hashCode()
        return hash * sourcePosition +
                sourcePosition * targetPosition +
                hash * targetPosition
    }

    override fun toString(): String {
        return "new " + javaClass.name + "<>(" + item.toString() + ", " + sourcePosition + ", " + targetPosition + ")"
    }

    private fun samePositions(change: ChangeOfPart<*>): Boolean {
        return sourcePosition == change.sourcePosition && targetPosition == change.targetPosition
    }

    override fun equals(other: Any?): Boolean {
        return other != null && other.javaClass == javaClass && samePositions(other as ChangeOfPart<*>)
    }

}

